pinot-hadoop/src/main/java/com/linkedin/pinot/hadoop/io/PinotRecordWriter.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.hadoop.io;
+
+import com.linkedin.pinot.common.utils.TarGzCompressionUtils;
+import com.linkedin.pinot.core.indexsegment.generator.SegmentGeneratorConfig;
+import com.linkedin.pinot.core.segment.creator.SegmentIndexCreationDriver;
+import com.linkedin.pinot.core.segment.creator.impl.SegmentIndexCreationDriverImpl;
+import com.linkedin.pinot.hadoop.job.JobConfigConstants;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * Basic Single Threaded {@link RecordWriter}
+ */
+public class PinotRecordWriter<K, V> extends RecordWriter<K, V> {
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(PinotRecordWriter.class);
+
+    private final SegmentGeneratorConfig _segmentConfig;
+    private final Path _workDir;
+    private final String _baseDataDir;
+    private PinotRecordSerialization _pinotRecordSerialization;
+    private final FileHandler _handler;
+    private long MAX_FILE_SIZE = 64 * 1000000L;
+
+    public PinotRecordWriter(SegmentGeneratorConfig segmentConfig, TaskAttemptContext context, Path workDir, PinotRecordSerialization pinotRecordSerialization) {
+        _segmentConfig = segmentConfig;
+        _workDir = workDir;
+        _baseDataDir = PinotOutputFormat.getTempSegmentDir(context) + "/data";
+        String filename = PinotOutputFormat.getTableName(context);
+        try {
+            _handler = new FileHandler(_baseDataDir, filename, ".json", MAX_FILE_SIZE);
```

Comment:
What happens when MAX_FILE_SIZE is not sufficient?

Reply:
Good one, The intermediate file writes similar to rolling logs. If the file size is more than `MAX_FILE_SIZE`, the FileHandler will flush the current file and create a new one.
https://github.com/linkedin/pinot/pull/2388/files#diff-917ef75d722ac12b40a9d4b36b5d3a3aR128
The intention behind the structure is
1. prevent a single large skew of the intermediate file
2. enable multi-threaded segment creation in next iteration